File: SorceryHex.Gba/pokemon_datatypes/maps.rb
mapPointers = types.FindMany "ppppwww", ->(b){
   b.Pointer "mapTileData", ->(b){
      b.Word "width" 
      b.Word "height" 
      b.Pointer "borderTile" 
      b.Pointer "tiles" 
      b.Pointer "tileset1" 
      b.Pointer "tileset2" 
      if types.Version == "BPRE" || types.Version == "BPGE" # FR / LG
        b.Byte "borderWidth"
        b.Byte "borderHeight"
        b.Unused 2
      end
   }
   b.Pointer "mapEventData", ->(b){
      persons = b.Byte "personCount"
      warps = b.Byte "warpCount"
      scripts = b.Byte "scriptCount"
      signposts = b.Byte "signpostCount"
      b.Array "persons", persons, ->(b){
         b.Byte "id"; b.Byte "picture"; b.Short "?"
         b.Short "x"; b.Short "y"
         b.Byte "?"; b.Byte "movementType"; b.Byte "movement"; b.Byte "?"
         b.Byte "isTrainer"; b.Byte "?"; b.Short "viewRadius"
         b.NullablePointer "script"
         b.Short "id"; b.Short "?"
      }
      b.Array "warps", warps, ->(b){
         b.Short "x"; b.Short "y"
         b.Byte "?"; b.Byte "warp"; b.Byte "map"; b.Byte "bank" # TODO link
      }
      b.Array "scripts", scripts, ->(b){
         b.Short "x"; b.Short "y"
         b.Short "?"; b.Short "scriptVariable"
         b.Short "scriptVariableValue"; b.Short "?"
         b.NullablePointer "script"
      }
      b.Array "signposts", signposts, ->(b){
         b.Short "x"; b.Short "y"
         b.Byte "talkingLevel"; b.Byte "signpostType"; b.Short "?"
         b.Word "?" # TODO script || -itemID .hiddenID .amount
      }
   }
   b.NullablePointer "script"
   b.NullablePointer "connections", ->(b){
      b.Word "count"
      b.NullablePointer "data"
   }
   b.Short "song"; b.Short "map"
   b.Byte "labelid"; b.Byte "flash"; b.Byte "weather"; b.Byte "type"
   b.Short "_"; b.Byte "labelToggle"; b.Byte "_"
}

bankPointers = types.FollowPointersUp mapPointers
leadPointer = types.FollowPointersUp bankPointers
self.maps = leadPointer[0].destination
self.mapdata = leadPointer[0].data

File: SorceryHex.Gba/pokemon_datatypes/maps.py
'''
static readonly Entry DataLayout = new Entry("map",
    new Entry("mapTileData", DataTypes.@pointer),
    new Entry("mapEventData",
        new Entry("personCount", DataTypes.@byte), new Entry("warpCount", DataTypes.@byte), new Entry("scriptCount", DataTypes.@byte), new Entry("signpostCount", DataTypes.@byte),
        new Entry("persons", "personCount",
            new Entry("?", DataTypes.@byte), new Entry("picture", DataTypes.@byte), new Entry("?", DataTypes.@byte), new Entry("?", DataTypes.@byte),
            new Entry("x", DataTypes.@short), new Entry("y", DataTypes.@short),
            new Entry("?", DataTypes.@byte), new Entry("movementType", DataTypes.@byte), new Entry("movement", DataTypes.@byte), new Entry("?", DataTypes.@byte),
            new Entry("isTrainer", DataTypes.@byte), new Entry("?", DataTypes.@byte), new Entry("viewRadius", DataTypes.@short),
            new Entry("script", DataTypes.@nullablepointer),
            new Entry("id", DataTypes.@short), new Entry("?", DataTypes.@byte), new Entry("?", DataTypes.@byte)
        ),
        new Entry("warps", "warpCount",
            new Entry("x", DataTypes.@short), new Entry("y", DataTypes.@short),
            new Entry("?", DataTypes.@byte), new Entry("warp", DataTypes.@byte), new Entry("map", DataTypes.@byte), new Entry("bank", DataTypes.@byte)
        ),
        new Entry("scripts", "scriptCount",
            new Entry("x", DataTypes.@short), new Entry("y", DataTypes.@short),
            new Entry("?", DataTypes.@short), new Entry("scriptVariable", DataTypes.@short),
            new Entry("scriptVariableValue", DataTypes.@short), new Entry("?", DataTypes.@short),
            new Entry("script", DataTypes.@nullablepointer)
        ),
        new Entry("signposts", "signpostCount",
            new Entry("x", DataTypes.@short), new Entry("y", DataTypes.@short),
            new Entry("talkingLevel", DataTypes.@byte), new Entry("signpostType", DataTypes.@byte), new Entry("?", DataTypes.@short),
            new Entry("?", DataTypes.@unknown4)
            // new Entry("dynamic", new DataType(4)) // *script || -itemID .hiddenID .amount || <missing>? //*/
        )
    ),
    new Entry("script", DataTypes.@nullablepointer),
    new Entry("connections", DataTypes.@nullablepointer,
        new Entry("count", DataTypes.@word),
        new Entry("data", DataTypes.@pointer)
    ),
    new Entry("song", DataTypes.@short), new Entry("map", DataTypes.@short),
    new Entry("label_id", DataTypes.@byte), new Entry("flash", DataTypes.@byte), new Entry("weather", DataTypes.@byte), new Entry("type", DataTypes.@byte),
    new Entry("_", DataTypes.@short), new Entry("labelToggle", DataTypes.@byte), new Entry("_", DataTypes.@byte)
);
'''

def mapLayout(b):
    b.ReadPointer("mapTileData", mapTileData)
    b.ReadPointer("mapEventData")
    b.ReadNullablePointer("script")
    b.ReadNullablePointer("connections")
    b.ReadShort("song")
    b.ReadShort("map")
    b.ReadByte("label_id")
    b.ReadByte("flash")
    b.ReadByte("weather")
    b.ReadByte("type")
    b.ReadShort("_")
    b.ReadByte("labelToggle")
    b.ReadByte("_")

def mapTileData(b):
    b.ReadWord("width")
    b.ReadWord("height")
    b.ReadPointer("borderTile")
    b.ReadPointer("tile")
    b.ReadPointer("tileset1")
    b.ReadPointer("tileset2")
    b.ReadByte("borderWidth")
    b.ReadByte("borderHeight")
    b.ReadShort("_")

matchingMapLayouts = types.FindMany(mapLayout)

# TODO do something with the matching map layouts. Find their root pointer or something.
